apps/src/music/views/ExemplarPlayerView.tsx

```diff
@@ -0,0 +1,152 @@
+import FontAwesomeV6Icon from '@code-dot-org/component-library/fontAwesomeV6Icon';
+import classNames from 'classnames';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+
+import {useAppDispatch, useAppSelector} from '@cdo/apps/util/reduxHooks';
+
+import {Source} from '../../lab2/types';
+import MusicBlocklyWorkspace from '../blockly/MusicBlocklyWorkspace';
+import {BlockMode, DEFAULT_PACK} from '../constants';
+import MusicLibrary from '../player/MusicLibrary';
+import MusicPlayer from '../player/MusicPlayer';
+import Simple2Sequencer from '../player/sequencer/Simple2Sequencer';
+import {setPlayerContext, setExemplarPlaybackEvents} from '../redux/musicRedux';
+
+import moduleStyles from './ExemplarPlayer.module.scss';
+
+interface ExemplarPlayerViewProps {
+  source: Source;
+  title: string;
+  labSetPlaying: (playing: boolean) => void;
+}
+
+const ExemplarPlayerView: React.FunctionComponent<ExemplarPlayerViewProps> = ({
+  source,
+  title,
+  labSetPlaying,
+}) => {
+  const dispatch = useAppDispatch();
+  const playerContext = useAppSelector(state => state.music.playerContext);
+
+  const playerRef = useRef<MusicPlayer | null>(null);
+  if (playerRef.current === null) {
+    playerRef.current = new MusicPlayer();
+  }
+  const workspaceRef = useRef<MusicBlocklyWorkspace>(
+    new MusicBlocklyWorkspace()
+  );
+  const simple2SequencerRef = useRef<Simple2Sequencer>(new Simple2Sequencer());
+  const [isPlaying, setIsPlaying] = useState<boolean>(false);
+  // Immediately load the library and source, then compile the song.
+  const onMount = useCallback(async () => {
+    workspaceRef.current.initHeadless();
+
+    workspaceRef.current.loadCode(source);
+    workspaceRef.current.compileSong(
+      {Sequencer: simple2SequencerRef.current},
+      BlockMode.SIMPLE2
+    );
+
+    // Clear any prior events and execute triggers to prepare playback.
+    simple2SequencerRef.current.clear();
+    workspaceRef.current.executeAllTriggers();
+    simple2SequencerRef.current.clear();
+    workspaceRef.current.executeCompiledSong();
+    const playbackEvents = simple2SequencerRef.current.getPlaybackEvents();
+    dispatch(setExemplarPlaybackEvents(playbackEvents));
+  }, [dispatch, source]);
+
+  useEffect(() => {
+    onMount();
+  }, [onMount]);
+
+  // Uses the already compiled song to preload sounds and play it.
+  const onPlaySong = useCallback(async () => {
+    Blockly.getMainWorkspace().hideChaff();
+    labSetPlaying(false);
+    playerRef.current?.stopSong();
+
+    const allTriggerEvents = simple2SequencerRef.current.getPlaybackEvents();
+
+    // Preload sounds and then play the song using the compiled events.
+    await playerRef.current?.preloadSounds(
+      [...allTriggerEvents, ...simple2SequencerRef.current.getPlaybackEvents()],
```

Comment:
I think this would be the same list twice, since we're not doing anything with the sequencer in between? `allTriggerEvents` should be equivalent to `simple2SequencerRef.current.getPlaybackEvents()`